public/js/user-profile.js

```diff
@@ -0,0 +1,181 @@
+const AvatarInput = document.getElementById("Upload-Avatar");
+const Pen1 = document.getElementById("Pen-1");
+const Pen2 = document.getElementById("Pen-2");
+const Pen3 = document.getElementById("Pen-3");
+const UserFeedbackFile = document.getElementById("Upload-Avatar-FileName");
+
+let userName = FullNameInput.value;
+let userEmail = EmailInput.value
+
+let errors = 0;
+const errorFlags = {
+	FullName: 1,
+	Email: 2,
+	Password: 4,
+	Avatar: 8
+}
+
+const getProfileDetails = async() => {
+	const response = await fetch('/profile-details');
+	const responseJson = await response.json();
+	FullNameInput.value = responseJson.name;
+	EmailInput.value = responseJson.email;
+	userName = FullNameInput.value;
+	userEmail = EmailInput.value;
+}
+
+const getAvatar = async() => {
+	const response = await fetch('/avatar')
+	try {
+		const responseJson = await response.json();
+		let base64 = responseJson.data;
+		base64 = `data:${responseJson.mimeType};base64,${base64}`;
+		document.getElementById("Base-Container-ProfilePicture").src = base64;
+	} catch(e) {
+		console.log(e);
+	}
+}
+
+async function executeUpdate(){
+	checkForInvalidInput(FullNameInput);
+	checkForInvalidInput(PasswordInput);
+	checkForInvalidInput(EmailInput);
+
+	const error = document.getElementById("Update-Error");
+	if(errors) {
+		error.innerText = "Please address above errors.";
+		error.style.color = "red";
+		return;
+	} else {
+		error.innerText = "";
+	}
+
+	const payload = {};
+	if(!FullNameInput.disabled) {
+		payload['name'] = FullNameInput.value;
+	}
+	if(!EmailInput.disabled) {
+		payload['email'] = EmailInput.value;
+	}
+	if(!PasswordInput.disabled) {
+		payload['pwd'] = PasswordInput.value;
+	}
+	if(AvatarInput.files[0]) {
+		// Make the POST call seperatly
+		const formData = new FormData();
+		formData.append('avatar', AvatarInput.files[0])
+		const response = await fetch("/avatar", {
+			method: "POST",
+			body: formData
+		})
+		const responseText = await response.text();
+		console.log(responseText);
+	}
+
+
+    const response = await fetch("/profile", {
+        method: "PATCH",
+        headers: {'Content-Type':'application/json'},
+        body: JSON.stringify(payload)
+    })
+
+    const ResponseText = await response.text(); 
+	const feedback = document.getElementById("Update-Error");
+    switch(ResponseText){
+        case "userUpdated":
+            // Successfully updated
+			feedback.innerText = "Successfully Updated Profile";
+			feedback.style.color = "green";
+            break;
+        case "emailInUse":
+            // Email already in use.
+			feedback.innerText = "That email is already in use";
+			feedback.style.color = "red";
+            break;
+        default:
+            feedback.innerText = "There was an issue on the server";
+			feedback.style.color = "red";
+    }
+
+	UserFeedbackFile.innerHTML = '';
+
+	getAvatar();
+	getProfileDetails();
+}
+
+function clearInput(x){
+    x.value = '';
+}
+
+function editFullName(){
```

Comment:
These `editFoo` functions could be generated by a higher order function, don't repeat yourself.

Reply:
E.g:

```js
function makeEditAction(inputEl, defaultValue){
    return function () {
        if (inputEl.disabled){
            inputEl.disabled= false; 
            inputEl.focus();
            inputEl.onfocus = clearInput(inputEl);
            return;
        }
        inputEl.disabled = true;
        inputEl.value = defaultValue;
    }
}

const editEmail = makeEditAction(EmailInput, userEmail);
const editFullName = makeEditAction(FullNameInput, userName);
const editPassword = makeEditAction(PasswordInput, "12345");
```